pythia/models/butd.py

```diff
@@ -108,52 +109,92 @@ def get_data_t(self, t, data, batch_size_t, prev_output):
+        return model_output
 
+    def forward_beam_search(self, sample_list, scores):
+        beam_search = BeamSearch(
+            self.vocab, self.config["inference"]["params"]["beam_length"]
+        )
+        sample_list = beam_search.init_batch(sample_list)
         batch_size = sample_list.image_feature_0.size(0)
         data, sample_list, timesteps = self.prepare_data(sample_list, batch_size)
         output = None
         batch_size_t = batch_size
         for t in range(timesteps):
             data, batch_size_t = self.get_data_t(t, data, batch_size_t, output)
-            if self.config["inference"]["type"] == "beam_search":
-                pi_t = data["texts"]
-            else:
-                pi_t = data["texts"][:, t].unsqueeze(-1)
+            pi_t = data["texts"]
             embedding = self.word_embedding(pi_t)
             attention_feature, _ = self.process_feature_embedding(
                 "image", sample_list, embedding[:, 0, :], batch_size_t=batch_size_t
             )
             output = self.classifier(attention_feature)
-
             # Compute Beam Search decoding
-            if self.config["inference"]["type"] == "beam_search":
-                finish, data, batch_size_t = beam_search.search(t, data, output)
-                if finish:
-                    break
-            else:
-                scores[:batch_size_t, t] = output
+            finish, data, batch_size_t = beam_search.search(t, data, output)
+            if finish:
+                break
 
         model_output = {"scores": scores}
-        if self.config["inference"]["type"] == "beam_search":
-            model_output["captions"] = beam_search.best_score()
+        model_output["captions"] = beam_search.best_score()
 
         return model_output
+
+    def forward_nucleus_sampling(self, sample_list, scores):
+        nucleus_sampling = NucleusSampling(
+            self.vocab, self.config["inference"]["params"]["sum_threshold"]
+        )
+        sample_list = nucleus_sampling.init_batch(sample_list)
+        batch_size = sample_list.image_feature_0.size(0)
+        data, sample_list, timesteps = self.prepare_data(sample_list, batch_size)
+        output = None
+        batch_size_t = batch_size
+        for t in range(timesteps):
+            data, batch_size_t = self.get_data_t(t, data, batch_size_t, output)
+            pi_t = data["texts"]
+            embedding = self.word_embedding(pi_t)
+            attention_feature, _ = self.process_feature_embedding(
+                "image", sample_list, embedding[:, 0, :], batch_size_t=batch_size_t
+            )
+            output = self.classifier(attention_feature)
+            # Compute Nucleus Sampling decoding
+            finish, data, batch_size_t = nucleus_sampling.sample(t, data, output)
+            if finish:
+                break
+
+        model_output = {"scores": scores}
+        model_output["captions"] = nucleus_sampling.get_caption()
```

Comment:
`get_caption` is used for nucleus_sampling, while `best_score` is used for beam search. It'd be nice to make the function name consistent across sampling strategies. I like `get_caption`.

Reply:
Agreed. Will make the required changes.

Reply:
With uniform structure this naming scheme will become consistent.